### PR TITLE
[BD-05][TNL-7946]: Waiting step details

### PR DESCRIPTION
**Pull request by giovannicimolin**
**TL;DR -** This PR implements the ORA waiting step details page and it's required APIs, along with some tooling to enable embedding react views inside ORA BackboneJS views.

This needs https://github.com/edx/edx-platform/pull/27032 to work properly.

JIRA: [TNL-7946](https://openedx.atlassian.net/browse/TNL-7946)

**What changed?**

There's a new screen in the instructor dashboard that helps identify learners stuck in the **waiting** step and manage their grades. 

The screen will show a list of students stuck in the `Waiting` step, and allow instructors to override grades and apply staff grading directly from that screen. 

Demo:
ora

**Developer Checklist**

- [ ] Reviewed the [release process](https://github.com/edx/edx-ora2/blob/master/.github/release_process.md)
- [ ] Translations and JS/SASS compiled
- [ ] Bumped version number in [setup.py](https://github.com/edx/edx-ora2/blob/a62e81a9b0d89223476967ec3c27f3557a850735/setup.py#L39) and [package.json](https://github.com/edx/edx-ora2/blob/a62e81a9b0d89223476967ec3c27f3557a850735/package.json#L3)

**Testing Instructions**

1. Checkout ORA and edx-platform related PRs.
2. Set up a ORA block with peer assessment (set the requirements to `1` grade and `1` peer review).
3. Log in as `honor`, submit the assessment.
4. Log in as `edx`, submit the assessment and assess `honor`'s submission. This will leave `edx` in the `waiting` position.
5. Go to the instructor dashboard, and click on the `waiting` column. 

**Reviewer Checklist**

Collectively, these should be completed by reviewers of this PR:

- [ ] I've done a visual code review
- [ ] I've tested the new functionality

FYI: @edx/masters-devs-gta

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @giovannicimolin! I've created [BLENDED-796](https://openedx.atlassian.net/browse/BLENDED-796) to keep track of it in Jira.   More details are on the [BD-05](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1529970851/BD-05+ORA+v3+Usability+Moderation) project page.


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Comment by giovannicimolin** [comment]
@muselesscreator FYI ^

File: openassessment/assessment/api/staff.py [hunk 1]
```diff
@@ -447,3 +447,30 @@ def _complete_assessment(
...
+    Args:
+        submission_uuid (str): The submission associated with this workflow.
+
+    Returns:
+        None
+
+    Raises:
+        StaffAssessmentInternalError: Raised when there is an internal error
+            creating the Workflow.
+
```
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply openassessment/assessment/api/staff.py:472]: Yup, thanks for catching it!

File: openassessment/assessment/models/staff.py [hunk 2]
```diff
@@ -148,6 +148,46 @@ def get_submission_for_review(cls, course_id, item_id, scorer_id):
...
+        Returns:
+            dict: a dictionary with the submission uuids as keys and their statuses as values.
+                  Example:
+                  {
+                      "uuid_1": "submitted",
+                      "uuid_2": "not_submitted
+                  }
+        """
+        # pylint: disable=unicode-format-string
+        timeout = (now() - cls.TIME_LIMIT).strftime("%Y-%m-%d %H:%M:%S")
```
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply openassessment/assessment/models/staff.py:169]: I've used it in a previous implementation, but yes, it's not being used now.

File: openassessment/xblock/static/js/src/lms/oa_staff_waiting_step.jsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    // and StaffAreaView `element` with our newly created one.
+    const staffAreaElement = $('.waiting-details-staff-area', reactElement);
+    this.baseView.element = staffAreaElement;
+    this.baseView.staffAreaView.element = staffAreaElement;
+    // Load staff area view
+    this.baseView.staffAreaView.load();
+  }
+}
+
+export default WaitingStepDetailsView;
```
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply openassessment/xblock/static/js/src/lms/oa_staff_waiting_step.jsx:54]: I've used the other backbone classes as a boilerplate for this but ended up heavily simplifying the implementation later on.
I'll turn this into a regular function.

File: openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+      setLoading(false);
+    }
+  };
+
+  useEffect(() => {
+    fetchWaitingStepDetails();
+
+    // Abort all requests on unmount
+    return function cancel() {
+      abortController.abort();
```
[the remark this answers is not in the document]
Comment (xitij2000) [reply openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx:39]: Or, you can also move this out all this code to the standard reducer-selector-thunk etc framework. However, given the small scope of this. My personal feeling is that this is a minor issue that can be ignored for now, and will actually disappear when more of ORA is built with react.

File: openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    const oraDescriptionText = gettext(
+      'The "{name}" problem is configured to require a minimum of {min_grades} '
+      + 'peer grades, and asks to review {min_graded} peers.',
+    ).replace(
+      '{name}', waitingStepDetails.display_name,
+    ).replace(
+      '{min_grades}', waitingStepDetails.must_be_graded_by,
+    ).replace(
+      '{min_graded}', waitingStepDetails.must_grade,
+    );
```
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx:53]: @xitij2000 This bit is JS code and the API doesn't send the fully translated string to the frontend.
I'll keep the current implementation since the rest of the JS code here also relies on it.

File: openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (giovannicimolin) [reply openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx:39]: @xitij2000 Thanks for the insights, I think that it's ok to ignore the issue for now. I've also removed the `abortController` since it wasn't actually doing anything regarding the issue.

**Comment by giovannicimolin** [comment]
@xitij2000 @jnlapierre Note for the reviewers here: I'll rebase the PR and squash some commits once this passes through a few review rounds.

I'm not rebasing now to avoid losing comments or context on fixes.

**Comment by jnlapierre** [comment]
Hello team :) Would it be possible to get the conflicts resolved before our code walkthrough huddle?

**Comment by giovannicimolin** [comment]
@jnlapierre The conflicts were mainly related to the version bump + static bundle compilation. I've rebased the PR to fix the conflicts.

File: openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+import WaitingStepList from './WaitingStepList';
+
+const WaitingStepPage = ({ waitingStepDataUrl }) => {
+  const [loading, setLoading] = useState(true);
+  const [error, setError] = useState(false);
+  const [waitingStepDetails, setWaitingStepDetails] = useState({
+    display_name: '', must_be_graded_by: '', must_grade: '', student_data: [],
+  });
+
+  const fetchWaitingStepDetails = async () => {
```
Comment (muselesscreator) [inline openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx:15]: Not a fan of network requests happening inside of component code...

File: openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    } finally {
+      setLoading(false);
+    }
+  };
+
+  useEffect(() => {
+    fetchWaitingStepDetails();
+  }, []);
+
+  const waitingStepPageContent = () => {
```
Comment (muselesscreator) [inline openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx:36]: This feels like it should be its own component

Comment (giovannicimolin) [reply openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx]: @muselesscreator Agreed, I've moved this to a separate component.

File: openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx (hunk 6 above)
Comment (giovannicimolin) [reply openassessment/xblock/static/js/src/lms/components/WaitingStepDetailsPage.jsx]: @muselesscreator I've pushed the request code to a file in the `api` folder.

**Comment by giovannicimolin** [comment]
@muselesscreator I think I've addressed all the points discussed in the meeting related to the frontend.

**Review by jnlapierre — APPROVED** [review]
LGTM This refactoring appears aligned with our discussion and review comments.

**Comment by giovannicimolin** [comment]
@jnlapierre I found a bug in the implementation. Don't merge this yet!

File: openassessment/assessment/models/staff.py [hunk 8]
```diff
@@ -148,6 +148,46 @@ def get_submission_for_review(cls, course_id, item_id, scorer_id):
...
 
+    @classmethod
+    def bulk_retrieve_workflow_status(cls, course_id, item_id, submission_uuids):
+        """
+        Retrieves a dictionary with the requested submission UUIDs statuses.
+
+        Args:
+            course_id (str): The course that this problem belongs to
+            item_ids (list of strings): The student_item (problem) that we want to know statistics about.
+
```
Comment (jnlapierre) [inline openassessment/assessment/models/staff.py:159]: Shouldn't submission_uuids also be defined here?

File: openassessment/assessment/models/peer.py [hunk 9]
```diff
@@ -248,6 +255,8 @@ def get_waiting_step_details(cls, course_id, item_id, submission_uuids):
         ).annotate(
             graded_count=models.Count('graded'),
             graded_by_count=models.Count('graded_by'),
+        ).filter(
+            graded_by_count__lt=must_be_graded_by
```
Comment (giovannicimolin) [inline openassessment/assessment/models/peer.py:259]: @jnlapierre This is the main backend change from the product review. This hides students that already received the required number of peer assessments from the waiting step dashboard since `grading_completed_at__isnull` doesn't always indicate that.

**Comment by marcotuts** [comment]
@jnlapierre  - all set on product review here (see also ora-dev slack channel ) but the updated messaging addresses the gap / issue we identified in the past week or two.

**Review by jnlapierre — APPROVED** [review]
LGTM
